### DISCUSSION TITLE
如何实现批注功能。

### DISCUSSION THREAD

**Discussion by dgsfor** [imzbf/md-editor-rt#79] [General]
我渲染了一篇markdown文档之后，想在网页上进行一些批注操作，请问这个有比较好的实现方案吗？
![image](https://github.com/imzbf/md-editor-rt/assets/22513159/851c2ff3-5a2a-4342-8627-5f44a5f24673)

**Comment by dgsfor** [comment]
类似于word的批注功能
![image](https://github.com/imzbf/md-editor-rt/assets/22513159/2b94581b-b07d-4aa7-90b1-efc3e0e81db2)

**Comment by imzbf** [comment]
这两个扩展可以实现类似的注释功能
1. https://github.com/markdown-it/markdown-it-footnote
2. https://github.com/markdown-it/markdown-it-abbr

编辑器扩展[参考](https://imzbf.github.io/md-editor-rt/en-US/docs#%F0%9F%92%B4%20Config%20Editor)

**Comment by dgsfor** [comment]
好的，不过我现在遇到了一个前置问题。是想在预览页面，对某段文字进行选择，然后可以弹出一个确认框：提示是否批注的字样。但是这个弹窗位置始终有点问题，不确定是否可以配置到编辑器里面。
1.比如我选择了这段文字，会弹出一个popconfirm框，这个逻辑是：
![image](https://github.com/imzbf/md-editor-rt/assets/22513159/6b56dc3c-0c77-4e29-837d-fb08bf915a9b)

![image](https://github.com/imzbf/md-editor-rt/assets/22513159/a56d5d8b-6b50-4409-aeea-04a3af23818f)
2.但是这个弹窗的位置，不知道应该如何做到内容里面，实现这样的效果
![image](https://github.com/imzbf/md-editor-rt/assets/22513159/ee0788a3-e2b9-4a96-8980-803cd95cdba4)
![image](https://github.com/imzbf/md-editor-rt/assets/22513159/a45aa91a-3b80-4cf6-b225-238c006d3961)

**Comment by dgsfor** [comment]
我还有想过监听鼠标右键，然后弹出功能菜单的方法，但是弹窗的位置始终不是在所选文字上方

**Comment by imzbf** [comment]
看起来位置没错，都是中间的，目前没有办法放到内部，文章内容是html文本不是各大框架的组件。想要指向选中的内容估计挺麻烦的，选中的内容并不是一个完整的html节点，绝对定到位当时鼠标所在位置比较靠谱

**Comment by dgsfor** [comment]
> 看起来位置没错，都是中间的，目前没有办法放到内部，文章内容是html文本不是各大框架的组件。想要指向选中的内容估计挺麻烦的，选中的内容并不是一个完整的html节点，绝对定到位当时鼠标所在位置比较靠谱

嗯 是的。我后面选择了另外一个方案来实现了。直接搞了一个弹窗，这个更简单了。
![image](https://github.com/imzbf/md-editor-rt/assets/22513159/887471db-1709-4148-a6ec-e17de5ba775a)
选择批注之后，会弹窗，然后填入批注内容。
![image](https://github.com/imzbf/md-editor-rt/assets/22513159/fa6ed006-aff9-45df-9e7b-70e00cf8e263)

针对这个，我有个问题想请教，如何实现内容高亮(就是我对某个文字进行选择批注，这段文字就会高亮。) 查了一下，有一个这个比较类似：https://github.com/alienzhou/web-highlighter/blob/master/README.zh_CN.md

**Comment by dgsfor** [comment]
```js
 function highlightABC(text, target) {
    const reg = new RegExp(`${target}`, 'gi');
    text = text.replace(reg, `<span style='background:red' color='black' size=4>${target}</span>`);
    return text;
  }
```
通过了一个笨办法来实现

**Comment by imzbf** [comment]
这个功能比较复杂，需要考虑挺多的，比如文章内容是否会更新，是否针对不同的用户表现不同？

像如果是个人文档，在这个编辑器里，基本上纯文本都有`data-line`的行号，你可以像你提供的那个站点一样，记录这个行号、选中内容开始、结束位置，以达到在渲染完成时重新去替换这部分内容，不过这种方式在内容有更新的时候可能就失效了。

那个站点好像是更新的源文本，就像把`markdown`的文本`这是一段高亮文字`更新为`这是一段<mark>高亮</mark>文字`。这倒是可以解决内容更新的时候，高亮内容不对的问题。

**Reply by dgsfor** [reply]
这个倒是一个好办法，昨天我测试实现了类似这样的批注功能，原理就是：替换这段文字加上类似<mark>标签，然后加上一个你之前提到的footnote。
![image](https://github.com/imzbf/md-editor-rt/assets/22513159/b96ad1f8-9014-4d49-9e55-ab44fa76e187)

这个批注我有两个想法：
- 第一是替换原文本内容，然后单独存储批注。
- 第二是不替换原文本内容，但是单独存储批注内容，并且记录原文本内容中批注的精确位置(这个也是我昨天纠结了半天，如何精确的定位这个文字内容，因为有重复的文本内容。)，最终在渲染的时候，先根据批注内容替换一下再展示。

**Reply by dgsfor** [reply]
这个在编辑器模式下确实有行号，貌似预览模式下没有吧，那我如何获取这个行号呢(问题还是在：如何精准定位我选择的批注位置)

**Reply by imzbf** [reply]
<img width="1016" alt="image" src="https://github.com/imzbf/md-editor-rt/assets/37092006/aaa2f63f-7d3b-4c7f-a3b4-635b52a3ea28">
都是有的，可以通过这个号码找到对应的markdown文本所在的内容的

**Reply by dgsfor** [reply]
看到了，确实有的。那我先按照这个方案来走走，感谢大佬~
